### ISSUE TITLE
crash with url returning status Code != 200

### ISSUE THREAD

**Issue by gustavopaes** [ampproject/ampbench#10]
Using "/api" route with an url that return statusCode != 200, the server crashes with the following error message:

```
[AMPBench:v.1.0][2017-02-04T13:01:20.863Z] [validator-signature:96c9247f6c997613302517643841f1f96e17bcf14ca3d51ebffae43b6274e648][HTTP:401] /api https://passwordprotected.com/amp.htm
_http_outgoing.js:356
    throw new Error('Can\'t set headers after they are sent.');
    ^

Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:356:11)
    at ServerResponse.header (/home/gpaes/git/ampbench/node_modules/express/lib/response.js:719:10)
    at ServerResponse.send (/home/gpaes/git/ampbench/node_modules/express/lib/response.js:164:12)
    at ServerResponse.json (/home/gpaes/git/ampbench/node_modules/express/lib/response.js:250:15)
    at on_output (/home/gpaes/git/ampbench/ampbench_routes.js:442:29)
    at IncomingMessage.res.on (/home/gpaes/git/ampbench/ampbench_lib.js:926:21)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
```

To fix it, it's necessary to prevent double `on_output_callback` call at `ampbench_lib.js`, at line [911](https://github.com/ampproject/ampbench/blob/master/ampbench_lib.js#L911) and [926](https://github.com/ampproject/ampbench/blob/master/ampbench_lib.js#L926). Maybe a return or else.

```
if (!http_response.statusIsOK()) { // NOT (200 == this.http_response_code)
    http_response.http_response_body = '';
    return on_output_callback(http_response, [CHECK_FAIL]); // !!! RETURN to front-end  - - - - - - - - - - - -
}
```

**Comment by pietergreyling** [comment]
@gustavopaes I am not seeing any server-side issues. 
When I try:
https://ampbench.appspot.com/validate?url=https://passwordprotected.com/amp.htm
I get:
```
AMP Checks [FAIL] [URL is unreachable] [Server SSL Certificate rejected[Error: self signed certificate[DEPTH_ZERO_SELF_SIGNED_CERT]]]
```
and with:
https://ampbench.appspot.com/api?url=https://passwordprotected.com/amp.htm
the result is as follows:
```
{
    "status": "FAIL",
    "url": "https://passwordprotected.com/amp.htm",
    "http_response": {
        "url": "https://passwordprotected.com/amp.htm",
        "http_code": 0,
        "http_text": "Server SSL Certificate rejected[Error: self signed certificate[DEPTH_ZERO_SELF_SIGNED_CERT]]",
        "response_time_ms": 0,
        "redirected": false,
        "redirects_count": 0,
        "redirects_urls": [],
        "is_https": true,
        "is_https_cert_authorized": false,
        "is_https_ssl_error": "Error: self signed certificate[DEPTH_ZERO_SELF_SIGNED_CERT]"
    },
    "amp_links": {
        "canonical_url": "",
        "amphtml_url": "",
        "amp_uses_feed": false
    },
    "results": []
}
```

**Comment by gustavopaes** [comment]
@pietergreyling sorry, I changed at the error message the original url request to this "passwordprotected.com" thinking this was not exist... but exists!

The problem is on /api and /api2 routes. And the page needs to return some statusCode >= 400, like these:

https://test-response-server.herokuapp.com/html/404
https://test-response-server.herokuapp.com/html/403
https://test-response-server.herokuapp.com/html/500

**Comment by nygellyndley** [comment]
Seeing the same issue.

Here's an example URL : https://ampbench.appspot.com/api?url=https%3A%2F%2Fampbench.appspot.com%2Fxxx

Repeat requesting this results in a 502 bad gateway from ampbench.appspot.com which is likely the app crashing and restarting.

**Comment by pietergreyling** [comment]
Tested it as follows and it does not crash:
[https://ampbench.appspot.com/api?url=https%3A%2F%2Fampbench.appspot.com%2Fxxx](https://ampbench.appspot.com/api?url=https%3A%2F%2Fampbench.appspot.com%2Fxxx)
returns this:
```
{
    "status": "FAIL",
    "url": "https://ampbench.appspot.com/xxx",
    "http_response": {
        "url": "https://ampbench.appspot.com/xxx",
        "http_code": 404,
        "http_text": "HTTP Status: 404 - Not Found",
        "response_time_ms": -1498490894790,
        "redirected": false,
        "redirects_count": 1,
        "redirects_urls": [
            "https://ampbench.appspot.com"
        ],
        "is_https": true,
        "is_https_cert_authorized": null,
        "is_https_ssl_error": ""
    },
    "amp_links": {
        "canonical_url": "",
        "amphtml_url": "",
        "amp_uses_feed": false
    },
    "results": []
}
```

And:
[https://ampbench.appspot.com/api?url=https://ampbench.appspot.com](https://ampbench.appspot.com/api?url=https://ampbench.appspot.com)
returns this:
```
{
    "status": "FAIL",
    "url": "https://ampbench.appspot.com",
    "http_response": {
        "url": "https://ampbench.appspot.com",
        "http_code": 200,
        "http_text": "HTTP Status: 200 - OK",
        "response_time_ms": 1,
        "redirected": false,
        "redirects_count": 1,
        "redirects_urls": [
            "https://ampbench.appspot.com"
        ],
        "is_https": true,
        "is_https_cert_authorized": true,
        "is_https_ssl_error": ""
    },
    "amp_links": {
        "canonical_url": "",
        "amphtml_url": "",
        "amp_uses_feed": false
    },
    "results": [
.....
    ]
}
```
Both are behaving correctly.

**Comment by gustavopaes** [comment]
Run it locally and you will see broking. At appspot the app restart after crash.

And the crash happens after response. That's is why you don't see any error at response.

**Comment by pietergreyling** [comment]
@gustavopaes 
FYI, please see this commit which should fix this issue: [Validator rollup + fix #10](https://github.com/ampproject/ampbench/commit/90a175f1f8486aac6cf1cd4c41ee293345a20d50)

**Comment by gustavopaes** [comment]
fixed!
